packages/flutter_tools/test/general.shard/runner/runner_test.dart

```diff
@@ -138,15 +191,31 @@ void main() {
 }
 
 class CrashingFlutterCommand extends FlutterCommand {
+  CrashingFlutterCommand({this.asyncCrash = false});
+
+  final bool asyncCrash;
+
   @override
   String get description => null;
 
   @override
-  String get name => 'test';
+  String get name => 'crash';
 
   @override
   Future<FlutterCommandResult> runCommand() async {
-    throw 'an exception % --'; // Test URL encoding.
+    const String error = 'an exception % --'; // Test URL encoding.
+    if (!asyncCrash) {
+      throw error;
+    }
+
+    final Completer<void> completer = Completer<void>();
+    Timer.run(() {
+      completer.complete();
+      throw error;
+    });
+
+    await completer.future;
+    return FlutterCommandResult.success();
```

Comment:
Does anything go wrong if you do `runCompleted.complete()` here instead?

Reply:
It seems to work, but it would be a bit more brittle.  If there asynchronous work ever occurs after `FlutterCommand.runCommand()` returns, the test could pass when it should fail.

Reply:
Sorry, I'm not quite getting it. Could you spell this out for me a bit more? What's the sequence of events that would be problematic?

Reply:
If `CrashingFlutterCommand.runCommand` completed `runCompleted` instead, the desired sequence of events would be:

1. `CrashingFlutterCommand.runCommand` generates an asynchronous error.
2. The `onError` callback from `runner.run`'s `runZoned` call fires.  We proceed down the crash reporting path. `WaitingCrashReporter` waits for `runCompleted`.
3. `CrashingFlutterCommand.runCommand` continues, completes `runCompleted`, and returns.
4. `runner.run` continues in its `try` block.
5. `WaitingCrashReporter` is unblocked and successfully reports the crash.

If execution yields between steps 3 and 4, then step 5 could run first.  If step 5 runs before step 4, a regression where the `try` block makes its own call to `exit()` would not be detected.

Reply:
Okay, thanks. I understand the concern now. Other than a stray call in `analyze_continuously.dart`, the tool only calls `exit()` in `runner.dart` and `base/signals.dart`. We should guard against proliferating calls to exit with a more on-the-nose test, but not in this PR. For this PR, I think it would be best to move the `runCompleted` completer from `runner.dart` into the test.

Reply:
I filed https://github.com/flutter/flutter/issues/59338